server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -340,222 +304,202 @@ public ClusterState execute(ClusterState currentState) {
-                        exceptions.add(
-                            ConfigurationUtils.newConfigurationException(processor.getType(), processor.getTag(), null, message)
-                        );
-                    }
-                }
-            }
-            ExceptionsHelper.rethrowAndSuppress(exceptions);
+    void validatePipeline(Map<DiscoveryNode, IngestInfo> ingestInfos, PutPipelineRequest request) throws Exception {
+        if (ingestInfos.isEmpty()) {
+            throw new IllegalStateException("Ingest info is empty");
         }
 
-        static ClusterState innerPut(PutPipelineRequest request, ClusterState currentState) {
-            IngestMetadata currentIngestMetadata = currentState.metaData().custom(IngestMetadata.TYPE);
-            Map<String, PipelineConfiguration> pipelines;
-            if (currentIngestMetadata != null) {
-                pipelines = new HashMap<>(currentIngestMetadata.getPipelines());
-            } else {
-                pipelines = new HashMap<>();
+        Map<String, Object> pipelineConfig = XContentHelper.convertToMap(request.getSource(), false, request.getXContentType()).v2();
+        Pipeline pipeline = Pipeline.create(request.getId(), pipelineConfig, processorFactories);
+        List<Exception> exceptions = new ArrayList<>();
+        for (Processor processor : pipeline.flattenAllProcessors()) {
+            for (Map.Entry<DiscoveryNode, IngestInfo> entry : ingestInfos.entrySet()) {
+                if (entry.getValue().containsProcessor(processor.getType()) == false) {
+                    String message = "Processor type [" + processor.getType() + "] is not installed on node [" + entry.getKey() + "]";
+                    exceptions.add(
+                        ConfigurationUtils.newConfigurationException(processor.getType(), processor.getTag(), null, message)
+                    );
+                }
             }
-
-            pipelines.put(request.getId(), new PipelineConfiguration(request.getId(), request.getSource(), request.getXContentType()));
-            ClusterState.Builder newState = ClusterState.builder(currentState);
-            newState.metaData(MetaData.builder(currentState.getMetaData())
-                .putCustom(IngestMetadata.TYPE, new IngestMetadata(pipelines))
-                .build());
-            return newState.build();
-        }
-
-        /**
-         * Returns the pipeline by the specified id
-         */
-        public Pipeline get(String id) {
-            return pipelines.get(id);
-        }
-
-        public Map<String, Processor.Factory> getProcessorFactories() {
-            return processorFactories;
         }
+        ExceptionsHelper.rethrowAndSuppress(exceptions);
     }
 
-    private static final class PipelineExecutionService {
-
-        private final PipelineStore store;
-        private final ThreadPool threadPool;
+    public void executeBulkRequest(Iterable<DocWriteRequest<?>> actionRequests,
```

Comment:
I don't think this method need to remain public?

Reply:
@martijnvg it does, it's called from `org.elasticsearch.action.bulk.TransportBulkAction#processBulkIndexIngestRequest`.

Reply:
ah, nevermind :)